client/src/features/Campaign/StepDialog.js

```diff
@@ -9,64 +9,196 @@ import {
+  }
+
+  const handleSave = () => {
+    setEditorTemplate({});
+    props.onSave();
+  }
+
+  const newTemplate = () => {
+    apiRequest('POST', '/templates', emptyTemplate)
+    .then( json => {
+      setEditorTemplate(json.template);
+      props.onNewTemplate(json.template);
+    })
+    .catch( e => {
+      console.log(e);
+    });
+  }
+
   return (
     <Dialog aria-labelledby="edit-step" 
             open={props.open}
-            onClose={props.onClose} >
-      <DialogContent>
+            onClose={props.onClose}
+            maxWidth="sm" >
+      <DialogContent className={classes.content}>
         <DialogTitle>
           {props.title}
         </DialogTitle>
-        <FormControl>
-          <InputLabel id="template-select-label">Template</InputLabel>
-          <Select labelId="template-select-label"
-                  id="template-select"
-                  value= {props.step ? props.step.templateId : null}
-                  onChange={e => props.step ? setTemplate(e) : props.setTemplateId(e.target.value)} 
-                  className={classes.select} >
-            {
-            props.templates.map( (template) => {
-              return (
-                <MenuItem key={template.id} value={template.id}>{template.name}</MenuItem>
-              )
-            })
-            }
-          </Select>
-        </FormControl>
+        <Grid container
+              alignItems="center"
+              justify="space-around" >
+          <Grid item xs={12} sm={9}>
+            <FormControl className={classes.fullWidth} variant="filled">
+              <InputLabel id="template-select-label">Template</InputLabel>
+              <Select labelId="template-select-label"
+                      id="template-select"
+                      value= {props.step ? props.step.templateId : null}
+                      onChange={e => props.step ? setTemplate(e) : props.setTemplateId(e.target.value)} 
+                      onChange={handleChange}
+                      className={classes.select} >
+                {
+                props.templates.map( (template) => {
+                  return (
+                    <MenuItem key={template.id} value={template.id}>{template.name}</MenuItem>
+                  )
+                })
+                }
+              </Select>
+            </FormControl>
+          </Grid>
+          <Grid item xs={12} sm={3}>
+            <Button variant="contained" 
+                    size="small" 
+                    color="primary" 
+                    onClick={newTemplate}
+                    className={classes.newButton} >
+              New
+            </Button>
+          </Grid>
+        </Grid>
+        <TemplateEditor template={editorTemplate}
+                        onSave={handleSaveTemplate} />
       </DialogContent>
       <DialogActions>
         { props.delete ? <Button onClick={props.onDeleteClick} color="secondary" className={classes.deleteButton}>Delete</Button> : null }
```

Comment:
you can use props.step instead to check I think

Reply:
I did try that originally but at certain points props.step becomes undefined so I needed a way to deal with that.